# Project 2: Data Import - Working with Web APIs and JSON (Movies Dataset)

# Project Brief for Self-Coders

Here you´ll have the opportunity to code major parts of Project 2 on your own. If you need any help or inspiration, have a look at the Videos or the Jupyter Notebook with the full code. <br> <br>
Keep in mind that it´s all about __getting the right results/conclusions__. It´s not about finding the identical code. Things can be coded in many different ways. Even if you come to the same conclusions, it´s very unlikely that we have the very same code. 

In [ ]:
import pandas as pd
import numpy as np
import json

## Importing Data from JSON files 

1. __Import__ the json files __blockbusters.json__, __blockbusters2.json__, __blockbusters3.json__ and load the datasets into Pandas DataFrames.


In [27]:
blockbuster = pd.read_json('blockbusters.json')
blockbuster1 = pd.read_json('blockbusters2.json')
movies = pd.read_json('movies.json')


In [11]:
blockbuster.sort_values(by='id').head(10)

,title,id,revenue,genres,belongs_to_collection,runtime
4,Titanic,597,1845034188,"[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...",None,194
11,Harry Potter and the Deathly Hallows: Part 2,12445,1341511219,"[{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...","{'id': 1241, 'name': 'Harry Potter Collection'...",130
1,Avatar,19995,2787965087,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...","{'id': 87096, 'name': 'Avatar Collection', 'po...",162
7,The Avengers,24428,1519557910,"[{'id': 878, 'name': 'Science Fiction'}, {'id'...","{'id': 86311, 'name': 'The Avengers Collection...",143
9,Avengers: Age of Ultron,99861,1405403694,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...","{'id': 86311, 'name': 'The Avengers Collection...",141
15,Frozen,109445,1274219009,"[{'id': 16, 'name': 'Animation'}, {'id': 12, '...","{'id': 386382, 'name': 'Frozen Collection', 'p...",102
5,Jurassic World,135397,1671713208,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...","{'id': 328, 'name': 'Jurassic Park Collection'...",124
2,Star Wars: The Force Awakens,140607,2068223624,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...","{'id': 10, 'name': 'Star Wars Collection', 'po...",136
8,Furious 7,168259,1506249360,"[{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...","{'id': 9485, 'name': 'The Fast and the Furious...",137
12,Star Wars: The Last Jedi,181808,1332539889,"[{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...","{'id': 10, 'name': 'Star Wars Collection', 'po...",152


In [10]:
movies.sort_values(by='id').head(10)

,title,id,revenue,genres,belongs_to_collection,runtime
2,Titanic,597,2264162353,"[{'id': 18, 'name': 'Drama'}, {'id': 10749, 'n...",None,194
12,Harry Potter and the Deathly Hallows: Part 2,12445,1341511219,"[{'id': 14, 'name': 'Fantasy'}, {'id': 12, 'na...","{'id': 1241, 'name': 'Harry Potter Collection'...",130
0,Avatar,19995,2923706026,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...","{'id': 87096, 'name': 'Avatar Collection', 'po...",162
7,The Avengers,24428,1518815515,"[{'id': 878, 'name': 'Science Fiction'}, {'id'...","{'id': 86311, 'name': 'The Avengers Collection...",143
10,Avengers: Age of Ultron,99861,1405403694,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...","{'id': 86311, 'name': 'The Avengers Collection...",141
15,Frozen,109445,1274219009,"[{'id': 16, 'name': 'Animation'}, {'id': 10751...","{'id': 386382, 'name': 'Frozen Collection', 'p...",102
5,Jurassic World,135397,1671537444,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...","{'id': 328, 'name': 'Jurassic Park Collection'...",124
3,Star Wars: The Force Awakens,140607,2068223624,"[{'id': 12, 'name': 'Adventure'}, {'id': 28, '...","{'id': 10, 'name': 'Star Wars Collection', 'po...",136
8,Furious 7,168259,1515400000,"[{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...","{'id': 9485, 'name': 'The Fast and the Furious...",137
13,Star Wars: The Last Jedi,181808,1332698830,"[{'id': 12, 'name': 'Adventure'}, {'id': 28, '...","{'id': 10, 'name': 'Star Wars Collection', 'po...",152


In [28]:
collection_df = pd.json_normalize(movies['belongs_to_collection'])

# Rename columns to avoid collisions (optional but good practice)
collection_df.columns = [f'collection_{col}' for col in collection_df.columns]

# Combine back with the original dataframe
movies = pd.concat([movies, collection_df], axis=1)

#genres
movies['genres'] = movies['genres'].apply(lambda x: [d['name'] for d in x] if isinstance(x, list) else [])


In [29]:
movies.head()

,title,id,revenue,genres,belongs_to_collection,runtime,collection_id,collection_name,collection_poster_path,collection_backdrop_path
0,Avatar,19995,2923706026,"[Action, Adventure, Fantasy, Science Fiction]","{'id': 87096, 'name': 'Avatar Collection', 'po...",162,87096.0,Avatar Collection,/uO2yU3QiGHvVp0L5e5IatTVRkYk.jpg,/gxnvX9kF7RRUQYvB52dMLPgeJkt.jpg
1,Avengers: Endgame,299534,2799439100,"[Adventure, Science Fiction, Action]","{'id': 86311, 'name': 'The Avengers Collection...",181,86311.0,The Avengers Collection,/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg,/zuW6fOiusv4X9nnW3paHGfXcSll.jpg
2,Titanic,597,2264162353,"[Drama, Romance]",None,194,NaN,NaN,NaN,NaN
3,Star Wars: The Force Awakens,140607,2068223624,"[Adventure, Action, Science Fiction]","{'id': 10, 'name': 'Star Wars Collection', 'po...",136,10.0,Star Wars Collection,/r8Ph5MYXL04Qzu4QBbq2KjqwtkQ.jpg,/zZDkgOmFMVYpGAkR9Tkxw0CRnxX.jpg
4,Avengers: Infinity War,299536,2052415039,"[Adventure, Action, Science Fiction]","{'id': 86311, 'name': 'The Avengers Collection...",149,86311.0,The Avengers Collection,/yFSIUVTCvgYrpalUktulvk3Gi5Y.jpg,/zuW6fOiusv4X9nnW3paHGfXcSll.jpg


## Working with APIs and JSON (Part 1)

2. __Create an account__ on https://www.themoviedb.org/

3. Get your personal __API Key__

4. __API-Request__ (movie module): Load all available information for the movie with __movie id = 140607__ into a Pandas DataFrame. <br> See https://developers.themoviedb.org/3/movies/get-movie-details for more information

## Working with APIs and JSON (Part 2)

5. __API-Request__ (discover module): Load all movies with __release date between 2020-01-01 and 2020-02-29__ into a Pandas DataFrame. <br>
See https://www.themoviedb.org/documentation/api/discover and https://developers.themoviedb.org/3/discover/movie-discover for more information.

##  Importing and Saving the Movies Dataset (Best Practice)

6. __API-Request__ (movie module): Load all available information for the movies with movie id = [__299534, 19995, 140607, 299536, 597, 135397, 420818, 24428, 168259, 99861, 284054, 12445, 181808, 330457, 351286, 109445, 321612, 260513__] into a Pandas DataFrame and __save the dataset in a local json file__.

# +++++++++ See some Hints below +++++++++++++

# ++++++++++++++ Hints +++++++++++++++++++++

__Hints for 1.__ <br>
To load json files you can use 

In [ ]:
with open("filename.json") as f:
    data = json.load(f)

and 

In [ ]:
pd.DataFrame(data), pd.read_json(filename.json), pd.json_normalize(data)

the json files have the following orientation (important when using pd.read_json()):
- blockbusters.json -> record
- blockbusters2.json -> column
- blockbusters3.json -> split 

__Hints for 4., 5., 6.__<br>
Make API GET-requests with the library requests (import requests):

In [ ]:
data = requests.get(url).json()

__Hints for 4. and 6.,__ <br> url structure for movie module:

"https://api.themoviedb.org/3/movie/insert_movie_id?api_key=insert_api_key" (replace "insert_movie_id" with movie id and "insert_api_key" with your personal api-key)

__Hints for 5.__<br>
url structure for discover module:

"https://api.themoviedb.org/3/discover/movie?api_key=insert_api_key&query1&query2..." (replace "insert_api_key" with your personal api-key and add appropriate queries)